shapely/geometry/geo.py

```diff
@@ -11,13 +11,44 @@
 from .collection import GeometryCollection
 
 
+def _is_coordinates_empty(coordinates):
+    """Helper to identify if list or all nested lists of coordinates are empty"""
+    if isinstance(coordinates, (list, tuple, np.ndarray)):
+        if len(coordinates) == 0:
+            return True
+        return all(map(_is_coordinates_empty, coordinates))
```

Comment:
Now that I think about it, the use of all() here makes this method relatively expensive. any(), used in the opposite sense, would return faster. What would you think about turning this into an is_coordinates_not_empty() function?

Reply:
Yeah, I had a similar thought. I thought it had to be done this way to check if _all_ the contents are empty. My first implementation just stopped on the first non-empty content. See this commit for example: https://github.com/Toblerity/Shapely/pull/852/commits/36abf1ab2132c844d4c02553d8d716538c7ed192

The issue I struggled with was what the shape should be if only one component was empty? For example, an empty point in a polygon?  Should it error out downstream as it does currently? 
```
>>> data = {"type": "Polygon", "coordinates": [[(), (1, 1), (1, 0)]]}
>>> geom = shape(data)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/shapely/geometry/geo.py", line 38, in shape
    return Polygon(ob["coordinates"][0], ob["coordinates"][1:])
  File "/usr/local/lib/python3.7/site-packages/shapely/geometry/polygon.py", line 242, in __init__
    ret = geos_polygon_from_py(shell, holes)
  File "/usr/local/lib/python3.7/site-packages/shapely/geometry/polygon.py", line 506, in geos_polygon_from_py
    ret = geos_linearring_from_py(shell)
  File "/usr/local/lib/python3.7/site-packages/shapely/geometry/polygon.py", line 442, in geos_linearring_from_py
    assert (n == 2 or n == 3)
AssertionError
```

Reply:
But, +1 to inverting the logic. I had the check of "stop when you hit one empty", but I can do the check of "stop when you hit one non-empty"

Reply:
@koshy1123 I had to take a few minutes to remind myself why fiona emits `[[]]` for the coordinates of a null geometry. It's by accident, not by design, that we see `[[]]` and not just `[]`. There's no special significance to the inner empty list and I think that in the future fiona may very well emit `[]` for all null/empty geometries or even just emit `None` for the entire geometry.

A mix of empty and filled coordinate pairs (or trios) within a coordinate sequence is not something we should try to smooth over. That's invalid data. I'm 100% sure that fiona and geopandas will never return such data.

This means that you can count on seeing `[]` in the case of empty/null points, linestrings, linearrings, or `[[]]` in the case of empty/null polygons, multipoints, multilinestrings, multipolygons, or geometry collections.

In light of this, I think the use of all() can remain. Does that makes sense?

Reply:
Got it 👍  Thanks for the deep dive! `all()` should be a fast compute in both then:
- if the coordinates field is non empty, the first non empty set will make it `False`
- if the coordinates field is empty, i.e, one of `[]` or `[[]]`  it's a fast compute to `True`